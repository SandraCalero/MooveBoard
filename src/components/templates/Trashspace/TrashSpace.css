.container {
	display: grid;
	grid-template-columns: repeat(auto-fill, minmax(20rem, 1fr));
	gap: 1rem;
	max-height: calc(100% - 15.3rem);
	overflow-y: scroll;
	justify-items: center;
	align-items: center;
}
