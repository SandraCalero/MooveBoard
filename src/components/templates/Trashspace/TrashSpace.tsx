import backToWorkspace from 'assets/icons/backToWorkspace.png';
import Button from 'components/atoms/Button/Button';
import Header from 'components/organisms/Header/Header';
import { Link } from 'react-router-dom';
import Modal from 'components/molecules/Modal/Modal';
import restorePostIt from 'assets/icons/restorePostIt.png';
import emptyTrashBin from 'assets/icons/emptyTrashBin.png';
import PostItList from 'components/molecules/PostItList/PostItList';

export default function TrashSpace() {
	return (
		<section>
			<Header>
				<Button
					variant='emptyTrash'
					text='Empty trash bin'
					icon={emptyTrashBin}
					altText='Empty trash bin'
				/>
				<Button
					variant='restorePostIts'
					text='Restore all post-its'
					icon={restorePostIt}
					altText='Restore all post-its'
				/>
				<Link to='/' className='link'>
					<img src={backToWorkspace} alt='Trash bin' />
					<span>Back to workspace </span>
				</Link>
			</Header>

			<PostItList
				postItListVariant='postItListDeleted'
				postItVariant='postItDeleted'
				disabled
			/>

			<Modal
				title='Delete Post It Note'
				message='Are you sure you want to delete this post it permanently?'
			/>
		</section>
	);
}
