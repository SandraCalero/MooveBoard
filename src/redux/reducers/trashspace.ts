/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IPostIt, IPostItEditable } from 'globals/definitions/postItProps';

interface trashspaceStateProps {
	deletedPostIts: Array<IPostIt>;
}

const localDeletedPostIts = localStorage.getItem('deletedPostIts');

const deletedPostIts: Array<IPostIt> = localDeletedPostIts
	? JSON.parse(localDeletedPostIts)
	: [];

const initialState: trashspaceStateProps = {
	deletedPostIts,
};

const trashspaceSlice = createSlice({
	name: 'trashspace',
	initialState,
	reducers: {
		addPostItToTrash: (state, action: PayloadAction<IPostIt>) => {
			state.deletedPostIts.unshift({
				id: action.payload.id,
				content: action.payload.content,
			});
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		addAllToTrash: (state, action: PayloadAction<Array<IPostItEditable>>) => {
			const newDeletedPostIts = action.payload.map((postIt) => ({
				id: postIt.id,
				content: postIt.content,
			}));
			state.deletedPostIts = [...state.deletedPostIts, ...newDeletedPostIts];
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		clearTrashBin: (state) => {
			state.deletedPostIts = [];
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		deletePostIt: (state, action: PayloadAction<number>) => {
			const postItIndex = state.deletedPostIts.findIndex(
				(postIt) => postIt.id === action.payload
			);
			state.deletedPostIts.splice(postItIndex, 1);
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
	},
});

export const { addPostItToTrash, addAllToTrash, clearTrashBin, deletePostIt } =
	trashspaceSlice.actions;

export default trashspaceSlice.reducer;
