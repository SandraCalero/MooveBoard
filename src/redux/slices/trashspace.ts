/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IPostIt } from 'globals/definitions/postItProps';

interface trashspaceStateProps {
	deletedPostIts: Array<IPostIt>;
}

const localDeletedPostIts = localStorage.getItem('deletedPostIts');

const deletedPostIts: Array<IPostIt> = localDeletedPostIts
	? JSON.parse(localDeletedPostIts)
	: [];

const initialState: trashspaceStateProps = {
	deletedPostIts,
};

const trashspaceSlice = createSlice({
	name: 'trashspace',
	initialState,
	reducers: {
		addPostItToTrash: (state, action: PayloadAction<IPostIt>) => {
			state.deletedPostIts.unshift({
				id: action.payload.id,
				content: action.payload.content,
			});
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		clearTrashBin: (state) => {
			state.deletedPostIts = [];
			localStorage.setItem(
				'deletedPostIts',
				JSON.stringify(state.deletedPostIts)
			);
		},
		deletePostIt: (state, action: PayloadAction<number>) => {
			console.log(state, action.payload);
		},
		restorePostIt: (state, action: PayloadAction<IPostIt>) => {
			console.log(state, action.payload);
		},
	},
});

export const { addPostItToTrash, clearTrashBin, deletePostIt, restorePostIt } =
	trashspaceSlice.actions;

export default trashspaceSlice.reducer;
