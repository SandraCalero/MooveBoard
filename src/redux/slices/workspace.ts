/* eslint-disable no-param-reassign */
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface PostItProps {
	id: number;
	variant: string;
	content?: string;
	draggable?: boolean;
	disabled?: boolean;
}

interface workspaceStateProps {
	currentPostItId: number;
	postItList: Array<PostItProps>;
}

const local = localStorage.getItem('postItList');
const postItList: Array<PostItProps> = local ? JSON.parse(local) : [];

const initialState: workspaceStateProps = {
	currentPostItId: 0,
	postItList,
};

const workspaceSlice = createSlice({
	name: 'workspace',
	initialState,
	reducers: {
		createPostIt: (state) => {
			state.currentPostItId += 1;
			state.postItList.push({
				id: state.currentPostItId,
				content: '',
				variant: 'postItCreated',
				draggable: true,
			});
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
		sendToTrash: (state, action) => {
			console.log('Send to trash', state, action.payload);
		},
		editPostIt(state, action: PayloadAction<PostItProps>) {
			state.postItList.forEach((postIt) => {
				if (postIt.id === action.payload.id) {
					postIt.content = action.payload.content;
				}
			});
			localStorage.setItem('postItList', JSON.stringify(state.postItList));
		},
	},
});

export const { createPostIt, sendToTrash, editPostIt } = workspaceSlice.actions;

export default workspaceSlice.reducer;
