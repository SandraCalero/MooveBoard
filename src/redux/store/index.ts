import { configureStore } from '@reduxjs/toolkit';
import workspaceReducer from 'redux/reducers/workspace';
import trashspaceReducer from 'redux/reducers/trashspace';

const store = configureStore({
	reducer: {
		workspace: workspaceReducer,
		trashspace: trashspaceReducer,
	},
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
