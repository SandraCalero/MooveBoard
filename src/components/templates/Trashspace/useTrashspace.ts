import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { clearTrashBin, deletePostIt } from 'redux/slices/trashspace';
import { RootState } from 'redux/store';
import { IPostIt } from 'globals/definitions/postItProps';
import { restoreAllPostIts } from 'redux/slices/workspace';

const useTrashspace = () => {
	const dispatch = useDispatch();
	const deletedPostIts = useSelector(
		(state: RootState) => state.trashspace.deletedPostIts
	);

	const initialPostIt: IPostIt = { id: 0, content: '' };

	const [postIt, setPostIt] = useState(initialPostIt);
	const [isModalOpen, setIsModalOpen] = useState(false);

	const shouldOpenModal = (postItInfo: IPostIt) => {
		setIsModalOpen(true);
		setPostIt(postItInfo);
	};

	const shouldCloseModal = () => {
		setIsModalOpen(false);
		setPostIt(initialPostIt);
	};

	const handleRestoreAll = () => {
		dispatch(clearTrashBin());
		dispatch(restoreAllPostIts(deletedPostIts));
	};

	const handleClearTrash = () => dispatch(clearTrashBin());

	const handleDeletePostIt = () => {
		dispatch(deletePostIt(postIt.id));
		shouldCloseModal();
	};

	return {
		shouldOpenModal,
		handleClearTrash,
		handleRestoreAll,
		shouldCloseModal,
		handleDeletePostIt,
		isModalOpen,
		deletedPostIts,
	};
};

export default useTrashspace;
