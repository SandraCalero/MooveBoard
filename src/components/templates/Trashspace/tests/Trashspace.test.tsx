import { BrowserRouter } from 'react-router-dom';
import { addPostItToTrash } from 'redux/slices/trashspace';
import store from 'redux/store';
import { render } from 'utils/tests/test-utils';
import Trashspace from '../Trashspace';

describe('<Trashspace />', () => {
	it('Snapshot', () => {
		store.dispatch(
			addPostItToTrash({
				id: 1,
				content: 'Testing add post it to trash bin action',
			})
		);
		const { container, getByText } = render(
			<BrowserRouter>
				<Trashspace />
			</BrowserRouter>
		);
		expect(container).toMatchSnapshot();
		expect(getByText('Testing add post it to trash bin action')).toBeTruthy();
	});
});
