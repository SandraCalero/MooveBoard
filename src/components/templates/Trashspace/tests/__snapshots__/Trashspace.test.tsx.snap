// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Trashspace /> should match the snapshot 1`] = `
<div>
  <main
    class="overflow-y-scroll h-full"
  >
    <header
      class="flex justify-around items-center p-3"
    >
      <button
        class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 w-20 h-28"
        name="button"
        type="button"
      >
        <img
          alt="Empty trash bin"
          class="h-4/5 w-4/5"
          src="emptyTrashBin.png"
        />
        Empty trash bin
      </button>
      <button
        class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 w-20 h-28"
        name="button"
        type="button"
      >
        <img
          alt="Restore all post-its"
          class="h-4/5 w-4/5"
          src="restoreAllPostIts.png"
        />
        Restore all post-its
      </button>
      <a
        class="w-20 h-28 flex flex-col items-center no-underline p-0 m-2 text-center hover:opacity-90 active:translate-y-1"
        href="/"
      >
        <img
          alt="Trash bin"
          class="w-4/5 h-4/5"
          src="backToWorkspace.png"
        />
        <span>
          Back to workspace 
        </span>
      </a>
    </header>
    <section>
      <h1
        class="m-2 text-center text-grey dark:text-mine-shaft-400 transition-colors"
      >
        Trash Bin
      </h1>
      <ul
        class="grid grid-cols-auto p-4 m-0 list-none gap-4 justify-center"
      >
        <li
          class="justify-self-center"
        >
          <div
            aria-hidden="true"
            class="flex flex-col items-center justify-center bg-yellow shadow-xl w-max dark:bg-cerulean-500 dark:shadow-lg dark:shadow-cerulean-700 transition-colors dark:border dark:border-grey dark:text-mine-shaft-600"
            id="1"
          >
            <div
              class="flex justify-between w-full px-2 pt-2 items-start"
            >
              <button
                class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 h-7 w-7 my-0 mx-0 opacity-20 hover:opacity-100 relative hover:after:content-[\\"Restore\\"] after:absolute after:left-8 after:bg-white after:rounded-lg after:p-1 after:text-sm dark:text-mine-shaft-900 dark:opacity-40 dark:hover:opacity-100 transition-colors"
                name="button"
                type="button"
              >
                <img
                  alt=""
                  class="h-4/5 w-4/5"
                  src="restorePostIt.png"
                />
              </button>
              <button
                class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 h-6 w-6 my-0 mx-0 opacity-20 text-xl hover:opacity-100 dark:text-mine-shaft-900 dark:opacity-40 dark:hover:opacity-100 transition-colors"
                name="button"
                type="button"
              >
                X
              </button>
            </div>
            <div
              class="break-words p-2 w-52 basis-52 bg-transparent"
            >
              Testing add post it to trash bin action
            </div>
          </div>
        </li>
      </ul>
      <div
        class="hidden items-center justify-center z-10 m-0 p-0 w-full
			h-full absolute inset-0 bg-transparent"
      >
        <div
          class="flex flex-col justify-center items-center bg-white max-w-xs h-auto rounded-lg shadow-2xl shadow-mine-shaft-900 dark:shadow-cerulean-700 dark:bg-cerulean-900 dark:border-none transition-colors"
        >
          <h2
            class="flex items-center p-3 m-2"
          >
            Delete Post It Note
          </h2>
          <div
            class="flex items-center p-3 m-0 border-y border-solid border-grey"
          >
            <img
              alt="Warning icon"
              class="m-2 w-10"
              src="warning.png"
            />
            <p
              class="m-2 text-sm "
            >
              Are you sure you want to delete this post it permanently?
            </p>
          </div>
          <div
            class="flex items-center p-3 m-0"
          >
            <button
              class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 justify-center w-24 h-12 rounded-lg bg-grey dark:bg-mine-shaft-500 dark:text-white transition-colors"
              name="button"
              type="button"
            >
              Cancel
            </button>
            <button
              class="flex flex-col items-center justify-center border-none
			m-2 p-0	bg-transparent hover:opacity-90 active:translate-y-1 justify-center w-24 h-12 rounded-lg bg-blue text-white dark:text-blue dark:bg-white font-semibold transition-colors"
              name="button"
              type="button"
            >
              Confirm
            </button>
          </div>
        </div>
      </div>
    </section>
  </main>
</div>
`;
